types/seamless-immutable/index.d.ts

```diff
@@ -91,29 +93,38 @@ declare namespace SeamlessImmutable {
+
             asObject<U extends object = {}, K extends keyof U = keyof U>(toKeyValue: (item: T) => [K, U[K]]): Immutable<U>;
             flatMap<TTarget>(mapFunction: (item: T) => TTarget): Immutable<TTarget extends any[] ? TTarget : TTarget[]>;
         }
 
         /** Custom implementation of the array functions, which return Immutable. */
         interface Overrides<T> {
-            map<TTarget>(mapFuction: (item: T) => TTarget): Immutable<TTarget[]>;
-            filter(filterFunction: (item: T) => boolean): Immutable<T[]>;
+            forEach(callbackfn: (value: Immutable<T>, index: number, array: Immutable<T[]>) => void, thisArg?: any): void;
+            map<TTarget>(mapFuction: (item: Immutable<T>) => TTarget): Immutable<TTarget[]>;
+            filter(filterFunction: (item: Immutable<T>) => boolean): Immutable<T[]>;
             slice(start?: number, end?: number): Immutable<T[]>;
-            concat(...arr: Array<T|T[]|Immutable<T|T[]>>): Immutable<T[]>;
-            reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T): Immutable<T>;
-            reduce<TTarget>(callbackfn: (previousValue: TTarget, currentValue: T, currentIndex: number, array: T[]) => TTarget, initialValue?: TTarget): Immutable<TTarget>;
-            reduceRight(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T): Immutable<T>;
-            reduceRight<TTarget>(callbackfn: (previousValue: TTarget, currentValue: T, currentIndex: number, array: T[]) => TTarget, initialValue?: TTarget): Immutable<TTarget>;
+            concat(...arr: Array<T|T[]|Immutable<T>|Array<Immutable<T>>|Immutable<T[]>>): Immutable<T[]>;
+            reduce(callbackfn: (previousValue: Immutable<T>, currentValue: Immutable<T>, currentIndex: number, array: Immutable<T[]>) => T): Immutable<T>;
+            reduce<TTarget>(callbackfn: (previousValue: TTarget, currentValue: Immutable<T>, currentIndex: number, array: Immutable<T[]>) => TTarget, initialValue?: TTarget): Immutable<TTarget>;
+            reduceRight(callbackfn: (previousValue: Immutable<T>, currentValue: Immutable<T>, currentIndex: number, array: Immutable<T[]>) => T): Immutable<T>;
+            reduceRight<TTarget>(callbackfn: (previousValue: TTarget, currentValue: Immutable<T>, currentIndex: number, array: Immutable<T[]>) => TTarget, initialValue?: TTarget): Immutable<TTarget>;
+        }
+
+        /** Merging this into Overrides breaks stuff, so this is split out */
+        interface ReadOnlyIndexer<T> {
+            readonly [key: number]: Immutable<T>;
         }
 
         /** These methods are banned by seamless-immutable. */
-        type MutatingArrayMethods = Extract<keyof any[], 'push' | 'pop' | 'sort' | 'splice' | 'shift' | 'unshift' | 'reverse'>;
+        type MutatingArrayMethods = Extract<keyof any[], 'push' | 'pop' | 'sort' | 'splice' | 'shift' | 'unshift' | 'reverse' | number>; // treating 'number' as a mutating method as it returns
```

Comment:
Hi @pheromonez,

Thank you for that. I am trying to remove the default indexer as it return a T when it should be an Immutable<T>. Any tips on what the correct thing would be to do to remove the default indexer (Is it even necessary when I'm replacing it)? I went with removing *number* as it seemed to give me the right results but I really wasn't sure. I'll give it another look over tonight.

Reply:
It looks like simply replacing the indexer as you've already done is sufficient.

Reply:
It looks like the removable of number is required. I had a bit of a play around:

This is what I see when I keep the original Exclude<... | number>:
![image](https://user-images.githubusercontent.com/11416577/57220170-a7150780-703d-11e9-9d06-9c110f6c01fa.png)

This is what happens when I do not exclude number:
![image](https://user-images.githubusercontent.com/11416577/57220182-b1370600-703d-11e9-8a72-d0fd8abd71a7.png)

Which seems to be giving Immutable<User> & User

Which is what I was expecting.
I believe this is the right check:
![image](https://user-images.githubusercontent.com/11416577/57220232-e04d7780-703d-11e9-97c8-b7c1929111a2.png)

I'm relying on Visual Studio to tell me as I wasn't sure how otherwise to check.

Reply:
You're right, at first the type definition for `MutatingArrayMethods` looked a little weird, but I makes sense to me now that we have to exclude the numbers as well. Good find!